[Blog post] - How Hugging Face Scaled Secrets Management for AI Infrastructure (#2657)
huggingface/blog: scaling-secrets-management.md

---
title: "How Hugging Face Scaled Secrets Management for AI Infrastructure" 
thumbnail: /blog/assets/infisical/thumbnail.png
authors:

> - user: guest

Reviewer: do you already have a infisical organization on the Hub?

(we can try to reclaim https://huggingface.co/infisical for you )

Reviewer: done, can you create an infisical organization now? https://huggingface.co/organizations/new

> - user: guest

Reviewer 2: ```suggestion
- user: segudev
  org: Infisical
```

---
title: "How Hugging Face Scaled Secrets Management for AI Infrastructure" 
thumbnail: /blog/assets/infisical/thumbnail.png
authors:

> - user: segudev
>   org: Infisical

Reviewer: ```suggestion
- user: segudev
  guest: true
  org: Infisical
```

---
title: "How Hugging Face Scaled Secrets Management for AI Infrastructure" 
thumbnail: /blog/assets/infisical/thumbnail.png
authors:
- user: segudev
  guest: true

>   org: Infisical

Reviewer: can you add a few team members to your org @thomas-infisical?

---

# How Hugging Face Scaled Secrets Management for AI Infrastructure

> Hugging Face has become synonymous with advancing AI at scale. With over 4 million builders deploying models on the Hub, the rapid growth of the platform necessitated a rethinking of how sensitive configuration data—secrets—are managed.

Reviewer 2: ```suggestion
Hugging Face has become synonymous with advancing AI at scale. With over 4 million builders deploying models on the Hub, the rapid growth of the platform necessitated a rethinking of how sensitive configuration data —secrets— are managed.
```

Last year, the engineering teams set out to improve the handling of their secrets and credentials. After evaluating tools like HashiCorp Vault, they ultimately chose [Infisical](https://infisical.com/).

This case study details their migration to Infisical, explains how they integrated its powerful features, and highlights how it enabled engineers to work more efficiently and securely.

## Background

> As Hugging Face's infrastructure evolved from an AWS-only setup to a multi-cloud environment that includes Azure and GCP, the engineering team needed a more agile, secure, and centralized way to manage secrets. Instead of reworking legacy systems or paying for heavyweight solutions like HashiCorp Vault, they turned to Infisical due to its developer-friendly workflows, multi-cloud abstraction, and robust security capabilities.

Reviewer 2: ```suggestion
As Hugging Face's infrastructure evolved from an AWS-only setup to a multi-cloud environment that includes Azure and GCP, the engineering team needed a more agile, secure, and centralized way to manage secrets. Instead of reworking legacy systems or adopting heavyweight solutions like HashiCorp Vault, they turned to Infisical due to its developer-friendly workflows, multi-cloud abstraction, and robust security capabilities.
```

The key challenges they faced were:

- An increased risk of “[secret sprawl](https://infisical.com/blog/what-is-secret-sprawl)” due to inconsistent management across environments.
- Complex permission management as the team scaled, requiring tight, role-based access controls (RBAC) integrated with the organization’s SSO (Okta).
- Difficulties with local development where traditional [.env files](https://infisical.com/blog/stop-using-env-files) compromised both security and developer productivity.

> - The burden of manual secret rotation, which became painfully evident after a security incident involving exposed credentials.

Reviewer 2: ```suggestion
- The burden of manual secret rotation, which became painfully evident after a security incident that involved exposed credentials.
```

In addition, the team needed a solution that adhered to infrastructure-as-code practices, supported project-by-project secret management, and provided a smooth balance between automation and manual control during deployments.

## Implementation

> Infisical’s flexible architecture was an ideal solution. The engineering team seized the opportunity to re-examine their internal project structure, splitting projects into distinct infrastructure and application domains. This allowed them to implement a clearer separation of concerns and standardize secret rotation practices—a priority in the wake of a recent security incident.

Reviewer 2: ```suggestion
Infisical’s flexible architecture was an ideal solution. The engineering team re-examined their internal project structure, splitting projects into distinct infrastructure and application domains. This allowed them to implement a clearer separation of concerns and standardize secret rotation practices—a priority in the wake of that recent security incident.
```

**Example:**

A new secret is required by an application running in Kubernetes. The secret can be created via the Infisical's CLI or the web UI, then the developer creates an `InfisicalSecret` resource in Kubernetes that specifies which secret from Infisical should be synced: 

```yaml
apiVersion: infisical.com/v1alpha1
kind: InfisicalSecret
metadata:
  name: my-app-secret
  namespace: production
spec:
  infisicalSecretId: "123e4567-e89b-12d3-a456-426614174000"
  targetSecretName: "my-app-k8s-secret"
```
Once the CRD is applied, the Infisical Operator continuously watches for updates. When changes are detected in Infisical, the Operator automatically updates the Kubernetes secret (`my-app-k8s-secret`).

```mermaid
graph TD
    A[Infisical Platform] -->|Push Update| B[Infisical Operator]
    B -->|Sync Secret| C[Kubernetes Secret (my-app-k8s-secret)]
    C -->|Mounted in| D[Application Pod]
    D -->|Reads| E[Environment Variables / Volumes]

> ```

Reviewer 2: I don't think this will be rendered, perhaps we need a link instead.

> Better yet, since the application's Deployment references `my-app-k8s-secret` as an environment variable source or mounted volume, the operator can automatically trigger a container reload when the the secret changes.

Reviewer 2: ```suggestion
Better yet, since the application's Deployment references `my-app-k8s-secret` as an environment variable source or mounted volume, the Operator can automatically trigger a container reload when the secret changes.
```

> In practice, Hugging Face engineers favor waiting for manual redeployments despite the operator’s ability to trigger container restarts automatically. This decision was driven by the need for precise control over deployments, particularly when high traffic (over 10 million requests per minute) and numerous replicas are involved.

Reviewer 2: ```suggestion
In practice, Hugging Face engineers favor waiting for manual redeployments despite the Operator’s ability to automatically trigger container restarts. This decision was driven by the need for precise control over deployments, particularly when high traffic (over 10 million requests per minute) and numerous replicas are involved.
```

Additionally, the [secret sharing](https://infisical.com/docs/documentation/platform/secret-sharing) functionality allows secure credentials sharing among ML/AI researchers at Hugging Face. The centralized Infisical platform also simplifies auditing and managing secret rotations—a necessity highlighted by previous security incidents.

## CI/CD and Infrastructure Integration

Seamless integration with CI/CD pipelines further enhanced the overall security posture. Infisical was embedded into the deployment pipeline via GitHub Actions using [OIDC authentication](https://infisical.com/docs/documentation/platform/identities/oidc-auth/github) and Terraform integration. By operating self-hosted runners within a secure environment, every deployment adhered to production-grade security standards. This integrated approach minimized risks and ensured a uniform experience from local development to cloud deployment.

## Technical Outcomes & Insights

Centralizing secrets management with Infisical brought tangible improvements:

> - ngineers no longer need to spend valuable time manually configuring environment secrets. Self-serve workflows accelerated onboarding and daily development cycles. 

Reviewer 3: Engineers no longer need to spend

- Automated audits and fine-grained access controls enabled rapid incident response and promoted a “shift left” approach to security.  
- Consistent integration across cloud providers, Kubernetes clusters, and CI/CD pipelines eliminated discrepancies in secret management, thus reinforcing the infrastructure's security and reliability.

As noted by Adrien Carreira, Head of Infrastructure at Hugging Face,

>"Infisical provided all the functionality and security settings we needed to boost our security posture and save engineering time. Whether you're working locally, running kubernetes clusters in production, or operating secrets within CI/CD pipelines, Infisical has a seamless prebuilt workflow."

## Conclusion

> Hugging Face's migration to Infisical demonstrates how a technically driven, engineering-centric approach to managing secrets across multiple cloud platforms delivers significants benefits. For tackling similar challenges, using Infisical is a practical way to work more efficiently while keeping security strong.

Reviewer 2: ```suggestion
Hugging Face's migration to Infisical demonstrates how a technically driven, engineering-centric approach to managing secrets across multiple cloud platforms delivers significant benefits. For tackling similar challenges, using Infisical is a practical way to work more efficiently while keeping security strong.
```

> When the secure path is made the easiest path, teams can focus on building innovative products instead of of worrying about managing secrets.

Reviewer 2: ```suggestion
When the secure path is made the easiest path, teams can focus on building innovative products instead of worrying about managing secrets.
```
